File: 第一阶段/1.5-字符串/StringOps.cs
// C# 字符串操作示例
// 对应文章：1.5 字符串操作

using System;

// ==================== 字符串创建 ====================
string s1 = "Hello";
string s2 = "World";
string s3 = @"C:\path\to\file";       // 原始字符串（@前缀）
string s4 = $@"Hello {s1}";           // 插值原始字符串

// ==================== 字符串拼接与插值 ====================
string result = s1 + " " + s2;         // 拼接
string interpolated = $"{s1} {s2}";    // 插值（推荐）

// ==================== 字符串格式化 ====================
Console.WriteLine("=== 字符串格式化 ===");
decimal price = 12.345m;
string name = "Alice";
decimal balance = 1234567.89m;

Console.WriteLine($"Price: {price:F2}");          // "Price: 12.35"
Console.WriteLine($"Name: {name,10}");            // "Name:      Alice"（右对齐）
Console.WriteLine($"Name: {-name,-10}");          // "Name: Alice     "（左对齐）
Console.WriteLine($"Balance: {balance:C}");       // "Balance: ¥1,234,567.89"（货币）
Console.WriteLine($"Hex: {255:X}");               // "Hex: FF"
Console.WriteLine($"Padded: {42:D5}");            // "Padded: 00042"

// ==================== 字符串访问与切片 ====================
Console.WriteLine("\n=== 字符串切片 ===");
string s = "Hello, World!";

// 单个字符访问
char first = s[0];               // 'H'
char last = s[s.Length - 1];     // '!'（C# 没有负数索引）

// C# 8.0+ 范围语法
string sub1 = s[..5];            // "Hello"（等同于 s[0:5]）
string sub2 = s[7..];            // "World!"（等同于 s[7:]）
string sub3 = s[..^1];           // "Hello, World"（去掉最后一个字符）
string sub5 = new string(s.Reverse().ToArray());  // 反转

Console.WriteLine($"原字符串:  {s}");
Console.WriteLine($"s[..5]:    {sub1}");
Console.WriteLine($"s[7..]:    {sub2}");
Console.WriteLine($"反转:      {sub5}");

// ==================== 常用字符串方法 ====================
Console.WriteLine("\n=== 常用方法 ===");
int index = s.IndexOf("World");          // 7
bool contains = s.Contains("Hello");     // true
bool starts = s.StartsWith("Hello");     // true
bool ends = s.EndsWith("World!");        // true

string replaced = s.Replace("World", "Python");  // "Hello, Python!"
string[] parts = s.Split(',');           // ["Hello", " World!"]
string joined = string.Join("-", new[] { "a", "b", "c" });  // "a-b-c"

Console.WriteLine($"IndexOf:  {index}");
Console.WriteLine($"Contains: {contains}");
Console.WriteLine($"Replace:  {replaced}");
Console.WriteLine($"Split:    [{string.Join(", ", parts)}]");
Console.WriteLine($"Join:     {joined}");

// 大小写与去空格
Console.WriteLine($"ToUpper:  {"hello".ToUpper()}");      // HELLO
Console.WriteLine($"ToLower:  {"HELLO".ToLower()}");      // hello
Console.WriteLine($"Trim:     {"  hello  ".Trim()}");     // hello

// ==================== 字符串不可变性 ====================
Console.WriteLine("\n=== 字符串不可变性 ===");
// C# 和 Python 的字符串都是不可变的！
// s[0] = 'h';  // 编译错误！

// 需要"修改"时，创建新字符串
string newS = "h" + s1[1..];     // "hello"
Console.WriteLine($"修改后: {newS}");

// 大量拼接用 StringBuilder（高效）
var sb = new System.Text.StringBuilder();
for (int i = 0; i < 5; i++)
{
    sb.Append(i);
    if (i < 4) sb.Append(",");
}
Console.WriteLine($"StringBuilder: {sb}");  // "0,1,2,3,4"

// ==================== 字符串比较 ====================
Console.WriteLine("\n=== 字符串比较 ===");
Console.WriteLine($"'abc' < 'abd':  {"abc" < "abd"}");     // True
Console.WriteLine($"'abc' == 'abc': {"abc" == "abc"}");    // True
Console.WriteLine($"'abc' != 'ABC': {"abc" != "ABC"}");    // True
// 注意：C# 默认区分大小写，Ordinal 比较

File: 第一阶段/1.5-字符串/string_ops.py
# Python 字符串操作示例
# 对应文章：1.5 字符串操作

# ==================== 字符串创建 ====================
s1 = "Hello"
s2 = "World"
s3 = r"C:\path\to\file"          # 原始字符串（不转义）
s4 = f"Hello {s1}"               # f-string（推荐，Python 3.6+）
s5 = """多行
字符串"""                          # 三引号多行字符串

# ==================== 字符串拼接与插值 ====================
result = s1 + " " + s2           # 拼接
interpolated = f"{s1} {s2}"      # f-string 插值（推荐）

# ==================== f-string 格式化（重点！） ====================
print("=== f-string 格式化 ===")
price = 12.345
name = "Alice"
balance = 1234567.89

# 基础格式化
print(f"Price: {price:.2f}")         # "Price: 12.35"（保留2位小数）
print(f"Name: {name:>10}")           # "Name:      Alice"（右对齐，宽度10）
print(f"Name: {name:<10}")           # "Name: Alice     "（左对齐）
print(f"Name: {name:^10}")           # "Name:   Alice   "（居中）
print(f"Name: {name:*^10}")          # "Name: ***Alice**"（居中填充*）
print(f"Balance: {balance:,.2f}")    # "Balance: 1,234,567.89"（千分位）
print(f"Hex: {255:#x}")             # "Hex: 0xff"（十六进制带前缀）
print(f"Percent: {0.756:.1%}")       # "Percent: 75.6%"（百分比）
print(f"Debug: {price=}")            # "Debug: price=12.345"（调试语法3.8+）

# 填充与对齐
print(f"{'left':<15}")              # "left           "
print(f"{'right':>15}")             # "          right"
print(f"{'center':^15}")            # "     center     "
print(f"{42:05d}")                  # "00042"（零填充）

# ==================== 字符串访问与切片 ====================
print("\n=== 字符串切片 ===")
s = "Hello, World!"

# 单个字符访问
first = s[0]        # 'H'
last = s[-1]        # '!'（负数索引！）

# 切片（Python 的杀手锏！）
sub1 = s[0:5]       # "Hello"
sub2 = s[7:]        # "World!"
sub3 = s[:5]        # "Hello"
sub4 = s[::2]       # "Hlo ol!"（每隔一个字符）
sub5 = s[::-1]      # "!dlroW ,olleH"（反转！C# 需要 ToCharArray() + Array.Reverse()）

print(f"原字符串:  {s}")
print(f"s[0:5]:    {sub1}")
print(f"s[7:]:     {sub2}")
print(f"s[::-1]:   {sub5}")

# ==================== 常用字符串方法 ====================
print("\n=== 常用方法 ===")
index = s.find("World")            # 7（找不到返回 -1）
contains = "Hello" in s            # True（用 in 运算符，不是方法！）
starts = s.startswith("Hello")     # True
ends = s.endswith("World!")        # True

replaced = s.replace("World", "Python")  # "Hello, Python!"
parts = s.split(',')              # ["Hello", " World!"]
joined = "-".join(["a", "b", "c"])  # "a-b-c"

print(f"find('World'):  {index}")
print(f"'Hello' in s:   {contains}")
print(f"replace:        {replaced}")
print(f"split:          {parts}")
print(f"join:           {joined}")

# 大小写与去空格
print(f"upper:  {'hello'.upper()}")       # HELLO
print(f"lower:  {'HELLO'.lower()}")       # hello
print(f"title:  {'hello world'.title()}") # Hello World
print(f"strip:  {'  hello  '.strip()}")   # hello（去两端空格）
print(f"lstrip: {'  hello  '.lstrip()}")  # hello  （去左空格）
print(f"rstrip: {'  hello  '.rstrip()}")  #   hello（去右空格）

# ==================== 字符串不可变性 ====================
print("\n=== 字符串不可变性 ===")
# 字符串创建后不能修改！
s = "Hello"
# s[0] = "h"  # TypeError! 字符串不支持单个字符赋值

# 需要"修改"时，创建新字符串
new_s = "h" + s[1:]    # "hello"
print(f"修改后: {new_s}")

# 大量拼接用 list + join（高效）
parts_list = []
for i in range(5):
    parts_list.append(str(i))
result = ",".join(parts_list)
print(f"join结果: {result}")  # "0,1,2,3,4"

# ==================== 字符串比较 ====================
print("\n=== 字符串比较 ===")
print(f"'abc' < 'abd':  {'abc' < 'abd'}")    # True（按字典序）
print(f"'abc' == 'abc': {'abc' == 'abc'}")   # True
print(f"'abc' != 'ABC': {'abc' != 'ABC'}")   # True（区分大小写）
